packages/js/product-editor/src/blocks/input/edit.tsx

```diff
@@ -0,0 +1,115 @@
+/**
+ * External dependencies
+ */
+import { createElement, createInterpolateElement } from '@wordpress/element';
+import type { BlockAttributes } from '@wordpress/blocks';
+import { useBlockProps } from '@wordpress/block-editor';
+import { useEntityProp } from '@wordpress/core-data';
+import { useInstanceId } from '@wordpress/compose';
+import { __ } from '@wordpress/i18n';
+import { Product } from '@woocommerce/data';
+import classNames from 'classnames';
+import {
+	BaseControl,
+	// @ts-expect-error `__experimentalInputControl` does exist.
+	__experimentalInputControl as InputControl,
+} from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { useValidation } from '../../contexts/validation-context';
+
+interface Metadata {
+	id?: number;
+	key: string;
+	value: string;
+}
+
+export function Edit( { attributes }: { attributes: BlockAttributes } ) {
+	const blockProps = useBlockProps();
+	const {
+		property,
+		label,
+		placeholder,
+		required,
+		validationRegex,
+		validationErrorMessage,
+	} = attributes;
+	const [ metadata, setMetadata ] = useEntityProp< Metadata[] >(
+		'postType',
+		'product',
+		'meta_data'
+	);
```

Comment:
Soon, the edited product will be in the context, and this will be the preferred way to access the edited product.

- https://github.com/woocommerce/woocommerce/issues/40233

Thinking about it, it might make it easier to support property and metadata once that is in place, by supplying a path...

- `editedProduct.name`
- `editedProduct.meta_data.custom_field`

I'll think about that as I implement the above issue (starting tomorrow).